runners/core-java/src/main/java/org/apache/beam/runners/core/metrics/ExecutionStateTracker.java

```diff
@@ -335,6 +375,12 @@ protected void takeSampleOnce(long millisSinceLastSample) {
       transitionsAtLastSample = transitionsAtThisSample;
     }
     updateMillisSinceLastTransition(millisSinceLastSample, state);
+    updateMillisSinceBundleStart(millisSinceLastSample);
+  }
+
+  @SuppressWarnings("NonAtomicVolatileUpdate")
+  private void updateMillisSinceBundleStart(long millisSinceLastSample) {
+    millisSinceBundleStart += millisSinceLastSample;
```

Comment:
I wanted to avoid the log construction which is going to log all the step name and processing times during every sample taken(very very high frequency compared to once in 20 mins when we are going to log). Hence, kept the check in DataflowExecutionStateTracker. I m open to design suggestions to implement this without constructing the log message.

Reply:
I think you could only construct the message when logging with my suggestion.
```
  private void updateMillisSinceBundleStart(long millisSinceLastSample) {
    // This variable is written by the Sampler thread, and read by the Progress Reporting thread.
    // Because only one thread modifies it, volatile provides enough synchronization.
    millisSinceLastStart += millisSinceLastSample;
    if (millisSinceLastStart > nextBundleLullReportMs) {
      // virtual reportBundleLull woudl construct the expensive message itself.
      reportBundleLull(trackedThread, millisSinceLastTransition);
      nextBundleLullReportMs += LULL_REPORT_MS;
    }
  }
```